File: Cpp/Example6-9-HillshadeMap/Example6-9-HillshadeMap.cpp
#include <gdal_priv.h>

#include <algorithm>
#include <array>
#include <iostream>
#include <vector>

using namespace std;

using F_RGB = std::array<double, 3>;

const double pi = 3.14159265358979323846;
const double d2r = pi / 180;
const double r2d = 180 / pi;

int demWidth;
int demHeight;
int dstBandNum = 4;

double geoTransform[6] = {0};
double startX;  //左上角点坐标X
double dx;      // X方向的分辨率
double startY;  //左上角点坐标Y
double dy;      // Y方向的分辨率

double noValue;

vector<float> demBuf;
vector<uint8_t> dstBuf;

//颜色查找表
vector<F_RGB> tableRGB(256);

//生成渐变色
void Gradient(F_RGB& start, F_RGB& end, vector<F_RGB>& RGBList) {
  F_RGB d;
  for (int i = 0; i < 3; i++) {
    d[i] = (end[i] - start[i]) / RGBList.size();
  }

  for (size_t i = 0; i < RGBList.size(); i++) {
    for (int j = 0; j < 3; j++) {
      RGBList[i][j] = start[j] + d[j] * i;
    }
  }
}

//初始化颜色查找表
void InitColorTable() {
  F_RGB blue({17, 60, 235});   //蓝色
  F_RGB green({17, 235, 86});  //绿色
  vector<F_RGB> RGBList(60);
  Gradient(blue, green, RGBList);
  for (int i = 0; i < 60; i++) {
    tableRGB[i] = RGBList[i];
  }

  F_RGB yellow({235, 173, 17});  //黄色
  RGBList.clear();
  RGBList.resize(60);
  Gradient(green, yellow, RGBList);
  for (int i = 0; i < 60; i++) {
    tableRGB[i + 60] = RGBList[i];
  }

  F_RGB red({235, 60, 17});  //红色
  RGBList.clear();
  RGBList.resize(60);
  Gradient(yellow, red, RGBList);
  for (int i = 0; i < 60; i++) {
    tableRGB[i + 120] = RGBList[i];
  }

  F_RGB white({235, 17, 235});  //紫色
  RGBList.clear();
  RGBList.resize(76);
  Gradient(red, white, RGBList);
  for (int i = 0; i < 76; i++) {
    tableRGB[i + 180] = RGBList[i];
  }
}

//根据高程选颜色
inline int GetColorIndex(double z, double min_z, double max_z) {
  int temp = (int)floor((z - min_z) * 255 / (max_z - min_z) + 0.6);
  return temp;
}

void ReadDem() {
  string workDir = getenv("GISBasic");
  string demPath = workDir + "/../Data/Terrain/dem.tif";

  GDALDataset* dem = (GDALDataset*)GDALOpen(demPath.c_str(), GA_ReadOnly);
  if (!dem) {
    cout << "Can't Open Image!" << endl;
    return;
  }

  demWidth = dem->GetRasterXSize();
  demHeight = dem->GetRasterYSize();

  dem->GetGeoTransform(geoTransform);
  startX = geoTransform[0];  //左上角点坐标X
  dx = geoTransform[1];      // X方向的分辨率
  startY = geoTransform[3];  //左上角点坐标Y
  dy = geoTransform[5];      // Y方向的分辨率

  noValue = dem->GetRasterBand(1)->GetNoDataValue();

  size_t demBufNum = (size_t)demWidth * demHeight;
  demBuf.resize(demBufNum, 0);

  int depth = sizeof(float);
  dem->GetRasterBand(1)->RasterIO(GF_Read, 0, 0, demWidth, demHeight,
                                  demBuf.data(), demWidth, demHeight,
                                  GDT_Float32, depth, demWidth * depth);

  GDALClose(dem);
  dem = nullptr;
}

// a b c
// d e f
// g h i
double CalHillshade(float* tmpBuf, double Zenith_rad, double Azimuth_rad,
                    double dx, double dy, double z_factor) {
  double dzdx = ((tmpBuf[2] + 2 * tmpBuf[5] + tmpBuf[8]) -
                 (tmpBuf[0] + 2 * tmpBuf[3] + tmpBuf[6])) /
                (8 * dx);
  double dzdy = ((tmpBuf[6] + 2 * tmpBuf[7] + tmpBuf[8]) -
                 (tmpBuf[0] + 2 * tmpBuf[1] + tmpBuf[2])) /
                (8 * dy);

  double Slope_rad = atan(z_factor * sqrt(dzdx * dzdx + dzdy * dzdy));
  double Aspect_rad = 0;
  if (abs(dzdx) > 1e-9) {
    Aspect_rad = atan2(dzdy, -dzdx);
    if (Aspect_rad < 0) {
      Aspect_rad = 2 * pi + Aspect_rad;
    }
  } else {
    if (dzdy > 0) {
      Aspect_rad = pi / 2;
    } else if (dzdy < 0) {
      Aspect_rad = 2 * pi - pi / 2;
    } else {
      Aspect_rad = Aspect_rad;
    }
  }

  double Hillshade =
      255.0 *
      ((cos(Zenith_rad) * cos(Slope_rad)) +
       (sin(Zenith_rad) * sin(Slope_rad) * cos(Azimuth_rad - Aspect_rad)));
  return Hillshade;
}

void Hillshade() {
  //设置方向：平行光
  double solarAltitude = 45.0;
  double solarAzimuth = 315.0;

  //
  double Zenith_rad = (90 - solarAltitude) * d2r;
  double Azimuth_math = 360.0 - solarAzimuth + 90;
  if (Azimuth_math >= 360.0) {
    Azimuth_math = Azimuth_math - 360.0;
  }
  double Azimuth_rad = (Azimuth_math)*d2r;

  size_t dstBufNum = (size_t)demWidth * demHeight * dstBandNum;
  dstBuf.resize(dstBufNum, 0);

  double z_factor = 1;
  double alpha = 0.3;  // A不透明度 α*A+(1-α)*B
  double minZ = *(std::min_element(demBuf.begin(), demBuf.end()));
  double maxZ = *(std::max_element(demBuf.begin(), demBuf.end()));

  // a b c
  // d e f
  // g h i
  for (int yi = 1; yi < demHeight - 1; yi++) {
    for (int xi = 1; xi < demWidth - 1; xi++) {
      size_t e = (size_t)demWidth * yi + xi;
      size_t f = e + 1;
      size_t d = e - 1;

      size_t b = e - demWidth;
      size_t c = b + 1;
      size_t a = b - 1;

      size_t h = e + demWidth;
      size_t i = h + 1;
      size_t g = h - 1;

      float tmpBuf[9] = {demBuf[a], demBuf[b], demBuf[c], demBuf[d], demBuf[e],
                         demBuf[f], demBuf[g], demBuf[h], demBuf[i]};
      double Hillshade =
          CalHillshade(tmpBuf, Zenith_rad, Azimuth_rad, dx, -dy, z_factor);

      size_t m = (size_t)demWidth * dstBandNum * yi + dstBandNum * xi;

      int index = GetColorIndex(demBuf[e], minZ, maxZ);

      for (int bi = 0; bi < 3; bi++) {
        double v = Hillshade * alpha + (1 - alpha) * (GByte)tableRGB[index][bi];
        dstBuf[m + bi] = (GByte)(std::min)((std::max)(v, 0.0), 255.0);
      }
      dstBuf[m + 3] = 255;
    }
  }
}

void WriteDst() {
  string workDir = getenv("GISBasic");
  string demPath = workDir + "/../Data/Terrain/dst.tif";

  GDALDriver* pDriver =
      GetGDALDriverManager()->GetDriverByName("GTIFF");  //图像驱动
  char** ppszOptions = NULL;
  ppszOptions =
      CSLSetNameValue(ppszOptions, "BIGTIFF", "IF_NEEDED");  //配置图像信息
  GDALDataset* dst = pDriver->Create(demPath.c_str(), demWidth, demHeight, 4,
                                     GDT_Byte, ppszOptions);
  if (!dst) {
    printf("Can't Write Image!");
    return;
  }

  dst->SetGeoTransform(geoTransform);

  int depth = sizeof(uint8_t);
  dst->RasterIO(GF_Write, 0, 0, demWidth, demHeight, dstBuf.data(), demWidth,
                demHeight, GDT_Byte, dstBandNum, nullptr, dstBandNum * depth,
                demWidth * dstBandNum * depth, depth);

  GDALClose(dst);
  dst = nullptr;
}

int main() {
  GDALAllRegister();  // GDAL所有操作都需要先注册格式

  //设置Proj数据
  std::string projDataPath = getenv("GISBasic");
  projDataPath += "/share/proj";
  CPLSetConfigOption("PROJ_LIB", projDataPath.c_str());

  InitColorTable();

  ReadDem();

  Hillshade();

  WriteDst();

  return 0;
}

File: Python/example-6-9-hillshade-map.py
import os
import sys
import math
import numpy as np
from osgeo import gdal

# =========================================================
# 常量
# =========================================================

PI = 3.14159265358979323846
D2R = PI / 180.0

# =========================================================
# 全局变量
# =========================================================

dem_width = 0
dem_height = 0

geo_transform = [0] * 6

start_x = 0.0
dx = 0.0

start_y = 0.0
dy = 0.0

no_value = 0.0

dem_buf = None
dst_buf = None

# 颜色查找表 LUT
table_rgb = np.zeros((256, 3), dtype=np.float32)

# =========================================================
# 颜色渐变
# =========================================================

def gradient(start, end, count):
    start = np.array(start, dtype=np.float32)
    end = np.array(end, dtype=np.float32)

    t = np.linspace(
        0.0,
        1.0,
        count,
        endpoint=False,
        dtype=np.float32
    ).reshape(-1, 1)

    return start + (end - start) * t


# =========================================================
# 初始化颜色表
# =========================================================

def init_color_table():
    global table_rgb

    blue = [17, 60, 235]
    green = [17, 235, 86]
    yellow = [235, 173, 17]
    red = [235, 60, 17]
    purple = [235, 17, 235]

    table_rgb[0:60] = gradient(blue, green, 60)
    table_rgb[60:120] = gradient(green, yellow, 60)
    table_rgb[120:180] = gradient(yellow, red, 60)
    table_rgb[180:256] = gradient(red, purple, 76)


# =========================================================
# 读取 DEM
# =========================================================

def read_dem():
    global dem_width
    global dem_height
    global geo_transform
    global start_x
    global dx
    global start_y
    global dy
    global no_value
    global dem_buf

    work_dir = os.getenv("GISBasicRepo")

    if not work_dir:
        print("错误：环境变量 GISBasicRepo 未设置。", file=sys.stderr)
        sys.exit(1)

    dem_path = os.path.join(
        work_dir,
        "Data",
        "Terrain",
        "dem.tif"
    )

    dem = gdal.Open(dem_path, gdal.GA_ReadOnly)

    if not dem:
        print("Can't Open Image!")
        sys.exit(1)

    dem_width = dem.RasterXSize
    dem_height = dem.RasterYSize

    geo_transform = dem.GetGeoTransform()

    start_x = geo_transform[0]
    dx = geo_transform[1]

    start_y = geo_transform[3]
    dy = geo_transform[5]

    band = dem.GetRasterBand(1)

    no_value = band.GetNoDataValue()

    # float32
    dem_buf = band.ReadAsArray().astype(np.float32)

    dem = None

    print(f"DEM Size: {dem_width} x {dem_height}")


# =========================================================
# Hillshade + 彩色渲染
# =========================================================

def cal_hillshade_vectorized(z_factor=1.0):

    global dst_buf

    # =====================================================
    # 光照参数
    # =====================================================

    solar_altitude = 45.0
    solar_azimuth = 315.0

    zenith_rad = (90.0 - solar_altitude) * D2R

    azimuth_math = 360.0 - solar_azimuth + 90.0

    if azimuth_math >= 360.0:
        azimuth_math -= 360.0

    azimuth_rad = azimuth_math * D2R

    # =====================================================
    # Horn 3x3
    #
    # a b c
    # d e f
    # g h i
    # =====================================================

    a = dem_buf[0:-2, 0:-2]
    b = dem_buf[0:-2, 1:-1]
    c = dem_buf[0:-2, 2:]

    d = dem_buf[1:-1, 0:-2]
    e = dem_buf[1:-1, 1:-1]
    f = dem_buf[1:-1, 2:]

    g = dem_buf[2:, 0:-2]
    h = dem_buf[2:, 1:-1]
    i = dem_buf[2:, 2:]

    # =====================================================
    # dzdx dzdy
    # =====================================================

    dzdx = (
        (c + 2.0 * f + i) -
        (a + 2.0 * d + g)
    ) / (8.0 * dx)

    dzdy = (
        (g + 2.0 * h + i) -
        (a + 2.0 * b + c)
    ) / (8.0 * abs(dy))

    # =====================================================
    # slope
    # =====================================================

    slope_rad = np.arctan(
        z_factor *
        np.sqrt(dzdx * dzdx + dzdy * dzdy)
    )

    # =====================================================
    # aspect
    # =====================================================

    aspect_rad = np.arctan2(dzdy, -dzdx)

    aspect_rad = np.where(
        aspect_rad < 0.0,
        2.0 * PI + aspect_rad,
        aspect_rad
    )

    mask = np.abs(dzdx) < 1e-9

    aspect_rad = np.where(
        mask,
        np.where(
            dzdy > 0.0,
            PI / 2.0,
            np.where(
                dzdy < 0.0,
                2.0 * PI - PI / 2.0,
                0.0
            )
        ),
        aspect_rad
    )

    # =====================================================
    # hillshade
    # =====================================================

    hillshade = 255.0 * (
        np.cos(zenith_rad) * np.cos(slope_rad) +
        np.sin(zenith_rad) *
        np.sin(slope_rad) *
        np.cos(azimuth_rad - aspect_rad)
    )

    hillshade = np.clip(
        hillshade,
        0.0,
        255.0
    )

    # =====================================================
    # 高程颜色 LUT
    # =====================================================

    valid_dem = dem_buf

    if no_value is not None:
        valid_dem = dem_buf[dem_buf != no_value]

    min_z = np.min(valid_dem)
    max_z = np.max(valid_dem)

    color_index = (
        (e - min_z) * 255.0 / (max_z - min_z) + 0.6
    ).astype(np.int32)

    color_index = np.clip(
        color_index,
        0,
        255
    )

    # LUT 查色
    dem_color = table_rgb[color_index]

    # =====================================================
    # 阴影混合
    # =====================================================

    alpha = 0.3

    hillshade_3 = hillshade[:, :, None]

    rgb = (
        hillshade_3 * alpha +
        dem_color * (1.0 - alpha)
    )

    rgb = np.clip(
        rgb,
        0.0,
        255.0
    ).astype(np.uint8)

    # =====================================================
    # RGBA 输出
    # =====================================================

    dst_buf = np.zeros(
        (dem_height, dem_width, 4),
        dtype=np.uint8
    )

    dst_buf[1:-1, 1:-1, 0:3] = rgb
    dst_buf[1:-1, 1:-1, 3] = 255

    # =====================================================
    # NoData 透明
    # =====================================================

    if no_value is not None:

        nodata_mask = (dem_buf == no_value)

        dst_buf[nodata_mask, 0] = 0
        dst_buf[nodata_mask, 1] = 0
        dst_buf[nodata_mask, 2] = 0
        dst_buf[nodata_mask, 3] = 0

    print("Hillshade Finished")


# =========================================================
# 输出 TIFF
# =========================================================

def write_dst():

    work_dir = os.getenv("GISBasicRepo")

    output_path = os.path.join(
        work_dir,
        "Data",
        "Terrain",
        "dst.tif"
    )

    driver = gdal.GetDriverByName("GTiff")

    dst = driver.Create(
        output_path,
        dem_width,
        dem_height,
        4,
        gdal.GDT_Byte,
        options=[
            'BIGTIFF=IF_NEEDED',
            'COMPRESS=LZW'
        ]
    )

    if not dst:
        print("Can't Write Image!")
        return

    dst.SetGeoTransform(geo_transform)

    # 如果原DEM有投影，继承
    work_dir = os.getenv("GISBasicRepo")

    dem_path = os.path.join(
        work_dir,
        "Data",
        "Terrain",
        "dem.tif"
    )

    src = gdal.Open(dem_path)

    if src:
        projection = src.GetProjection()

        if projection:
            dst.SetProjection(projection)

        src = None

    # 写RGBA
    for i in range(4):
        band = dst.GetRasterBand(i + 1)

        band.WriteArray(dst_buf[:, :, i])

    dst.FlushCache()

    dst = None

    print("彩色 Hillshade 已保存:")
    print(output_path)


# =========================================================
# 自动设置 PROJ_LIB
# =========================================================

def setup_proj_lib():

    if 'CONDA_PREFIX' in os.environ:

        conda_prefix = os.environ['CONDA_PREFIX']

        # Windows
        proj_path = os.path.join(
            conda_prefix,
            'Library',
            'share',
            'proj'
        )

        if not os.path.exists(proj_path):

            # Linux/macOS
            proj_path = os.path.join(
                conda_prefix,
                'share',
                'proj'
            )

        if os.path.exists(proj_path):

            os.environ['PROJ_LIB'] = proj_path

            print("PROJ_LIB =", proj_path)

            return True

    return False


# =========================================================
# Main
# =========================================================

if __name__ == "__main__":

    gdal.AllRegister()

    setup_proj_lib()

    init_color_table()

    read_dem()

    cal_hillshade_vectorized()

    write_dst()